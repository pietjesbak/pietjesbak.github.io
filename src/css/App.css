.container {
    display: flex;
    max-width: 1100px;
    margin: 0 auto;
    padding-top: 30px;
}

.wrapper {
    max-width: 1100px;
    margin: 0 auto;
    margin-bottom: 10px;
}

.page-wrap {
    min-height: 100%;
    margin-bottom: -100px;
    /* equal to footer height */
}

.page-wrap:after {
    content: "";
    display: block;
}

footer, .page-wrap:after {
    height: 100px;
}

footer {
    background: #103D5D;
    color: #fff;
}

footer ul {
    list-style: none;
    margin: 0;
}

footer ul>li {
    display: inline-block;
    width: 115px;
    padding: 10px;
    margin: 10px 0;
    background: rgb(76, 109, 128);
}

footer ul>li:first-child {
    margin-right: 10px
}

footer ul>li:hover {
    background: rgb(93, 123, 139);
}

footer ul>li a {
    color: #fff!important;
    text-decoration: none;
}

footer>span {
    display: inline-block;
    text-align: center;
    width: 100%;
}

nav.menu ul {
    list-style: none;
    padding: 0;
    margin: 20px 5px;
}

nav.menu ul li {
    display: inline-block;
}

nav.menu ul li:not(:last-child) {
    margin-right: 10px;
}

nav.menu ul li.right {
    float: right;
}

nav.menu ul li.right.small {
    width: 40px;
}

button, .button, .small-button {
    display: inline-block;
    text-align: center;
    text-decoration: none;
    border-radius: 2px;
    background-color: #FA6900;
    cursor: pointer;
    border: 0;
    min-width: 120px;
    color: #fff!important;
    font-size: 14px;
    box-shadow: 0 2px 5px 0 rgba(0, 0, 0, .26);
    padding: 10px 6px;
}

button:hover, .button:hover, .small-button:hover, .button.active {
    position: relative;
    top: 2px;
    box-shadow: 0 -2px 5px 0 rgba(0, 0, 0, .26);
}

.small-button {
    min-width: 40px;
    width: 40px;
}

.user-profile {
    position: fixed;
    bottom: 20px;
    right: 40px;
    width: 100px;
    height: 100px;
    z-index: 999;
}

.user-profile img {
    border-radius: 50%;
    max-width: 100%;
}

.location {
    padding: 10px 5px;
    background: #fff;
    margin: 10px 0;
    display: inline-block;
    border-radius: 5px;
}
