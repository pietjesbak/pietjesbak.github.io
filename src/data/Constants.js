/**
 * A heroku app running cors-anywhere to access some boardgame geek api endpoints.
 * https://github.com/Rob--W/cors-anywhere
 *
 * @const
 * @type {string}
 */
export const CORS_ANYWHERE_DYNO = 'https://pietjesbak.herokuapp.com/';

/**
 * The pietjesbak game collection on board game geek.
 *
 * @const
 * @type {string}
 */
export const PIETJESBAK_BBG_COLLECTION = 'https://www.boardgamegeek.com/xmlapi/collection/de%20pietjesbak?own=1';

/**
 * Board game geek game details api. Add the board game id. Requires the cors-anywhere dyno.
 *
 * @const
 * @type {string}
 */
export const BBG_GAME_API = 'https://www.boardgamegeek.com/xmlapi/boardgame/';

/**
 * Facebook pietjesbak events graph api.
 * Please don't steal my app secret k thanks.
 *
 * @const
 * @type {string}
 */
export const FACEBOOK_PIETJESBAK_EVENTS = 'https://graph.facebook.com/v2.10/gezelschapsspellenpietjesbak/events?access_token=874889636002007|788af4e5d543244dff7ee209987a2a57&format=json&limit=1';
