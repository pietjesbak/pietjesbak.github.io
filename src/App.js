import './css/App.css';
import { auth } from './Firebase.js';
import { HashRouter, Route, Switch } from 'react-router-dom';
import BggList from './BggList.js';
import BggRequests from './BggRequests.js';
import Facebook from './Facebook.js';
import IconButton from './IconButton.js';
import IconLink from './IconLink.js';
import inventory from './data/Inventory.js';
import React, { Component } from 'react';
import SimpleMap from './SimpleMap.js';
import Tooltip from 'react-simple-tooltip';
import ScrollToTop from './ScrollToTop';

class App extends Component {
    constructor() {
        super();
        this.state = {
            username: ''
        };
    }

    componentDidMount() {
        auth.onAuthStateChanged(user => {
            if (user) {
                this.setState({ username: user.displayName });
            }
        });
    }

    logout = async () => {
        await inventory.logout();
        this.setState({
            nameuser: ''
        });
    }

    login = async () => {
        const user = await inventory.login();
        this.setState({
            username: user.displayName
        });
    }

    home() {
        return (
            <div>
                <Facebook></Facebook>
                <BggRequests></BggRequests>
            </div>
        );
    }

    info() {
        return (
            <div>
                <div className="card">
                    <h3>Wie zijn wij?</h3>
                    Spellenclub De pietjesbak is een spellenclub in Gent / Mariakerke. <br />
                    Elke derde vrijdag van de maand (behalve in juli en augustus) <br /><br />

                    Voor 1 € inkom is iedereen welkom om een nieuw of oud spel mee te komen spelen. <br />
                    Oud of Jong, moeilijk of makkelijk, we brengen steeds een grote variatie aan spellen mee die ter plaatse uitgelegd worden. <br />
                    We starten rond 19u30 en gaan door tot ± 23 a 24u.

                    <h3>In 't Geestje</h3>
                    <span className="location">
                        Zandloperstraat 83 <br />
                        9030 Mariakerke
                    </span>
                    <SimpleMap></SimpleMap>
                </div>
            </div>
        );
    }

    noRoute() {
        return (
            <div className="center not-found">
                <span>404</span>
                <div style={{ fontSize: 14 + "vw", whiteSpace: "nowrap" }}>¯\_(ツ)_/¯</div>
            </div>
        )
    }

    renderLoginButton() {
        if (inventory.user !== null) {
            return <IconButton subClass="logout" icon="logout" text="Log uit" placement="left" action={this.logout}></IconButton>
        } else {
            return <IconButton subClass="login" icon="login" text="Log in" placement="left" action={this.login}></IconButton>
        }
    }

    render() {
        return (
            <HashRouter className="app">
                <ScrollToTop>
                    <div className="full-height">
                        <div className="page-wrap">
                            <header>
                                <div className="wrapper">
                                    <h1>
                                        <svg id="logo" width="57" height="52">
                                            <path d="M4 25.650635094610966L16.5 4L41.5 4L54 25.650635094610966L41.5 47.30127018922193L16.5 47.30127018922193Z" strokeWidth="5" fill="#c33" stroke="#eee"></path>
                                        </svg>
                                        Spellenclub De Pietjesbak
                                </h1>
                                </div>
                            </header>
                            <div className="wrapper">
                                <nav className="menu">
                                    <ul>
                                        <li><IconLink icon="globe" to="/" text="Home"></IconLink></li>
                                        <li><IconLink icon="megaphone" to="/info" text="Info"></IconLink></li>
                                        <li><IconLink icon="puzzle" to="/games" text="Games"></IconLink></li>
                                        <li className="right">
                                            {this.renderLoginButton()}
                                        </li>
                                    </ul>
                                </nav>
                                {inventory.user !== null ?
                                    <div>
                                        <div className="user-profile">
                                            <Tooltip content={`Ingeloged als ${this.state.username}`} placement="left">
                                                <img alt={this.state.username} src={inventory.user.photoURL} />
                                            </Tooltip>
                                        </div>
                                    </div>
                                    :
                                    <div></div>
                                }
                                <Switch>
                                    <Route exact path="/" component={this.home} />
                                    <Route path="/games" render={() => <BggList></BggList>} />
                                    <Route path="/info" component={this.info} />
                                    <Route render={this.noRoute} />
                                </Switch>
                            </div>
                        </div>
                        <footer>
                            <ul className="wrapper">
                                <li>
                                    <a href="https://www.facebook.com/gezelschapsspellenpietjesbak/">
                                        <i className="icon-facebook-squared"></i>
                                        Like
                                </a>
                                </li>
                                <li>
                                    <a href="mailto:pietdecoensel@gmail.com">
                                        <i className="icon-mail-alt"></i>
                                        Contact
                                </a>
                                </li>
                            </ul>
                            <span>&copy; Spellenclub De Pietjesbak &#9679; Lennert Claeys</span>
                        </footer>
                    </div>
                </ScrollToTop>
            </HashRouter>
        );
    }
}

export default App;
