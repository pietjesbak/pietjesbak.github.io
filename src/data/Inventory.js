import { PIETJESBAK_BBG_COLLECTION } from './Constants';
import * as data from './BggData.js';
import firebase, { auth, provider } from '../Firebase.js';
import * as constants from './Constants.js';

class Inventory {
    constructor() {
        /**
         * A promise for all board game geek games.
         *
         * @private
         * @type {?Promise.<Map.<number, BggGameData>>}
         */
        this.fetchGamePromise_ = null;

        /**
         * A map of games you requested and their firebase key.
         *
         * @private
         * @type {?Map.<number, string>}
         */
        this.ownRequestedGames_ = null;

        /**
         * A map of all requested games and the amount of times they were requested this month.
         *
         * @private
         * @type {?Map.<number, number>}
         */
        this.requestedGames_ = null;

        /**
         * An array of callbacks for when the game lists update.
         *
         * @private
         * @type {Array.<function(Map.<number, BggGameData>)>}
         */
        this.changeListeners_ = [];

        /**
         * Your user object when logged in to firebase.
         *
         * @private
         * @type {?Object}
         */
        this.user_ = null;

        /**
         * All users of the pietjesbak. uid => name
         *
         * @private
         * @type {Map.<string, string>}
         */
        this.users_ = new Map();

        this.firebaseWaitForAuthChange_();
        this.getFirebaseRequests_();
        // this.getFirebaseUsers_();
    }

    /**
     * Get the board game collection from boardgame geek.
     *
     * @return {Promise.<Map.<number, BggGameDate>>}
     */
    async getGames() {
        if (this.fetchGamePromise_ === null) {
            this.fetchGamePromise_ = this.fetchGames_();
        }

        return await this.fetchGamePromise_;
    }

    /**
     * Toggles a request for a game. You need to be logged in for this.
     *
     * @param {*} game The game to toggle.
     * @return {Promise}
     */
    async toggleGame(game) {
        const date = new Date();
        const key = this.ownRequestedGames_.get(game.id);

        if (this.user_ === null) {
            throw new Error('You need to be logged in to request a game.');
        }

        if (key === undefined) {
            // Request it.
            await firebase.database().ref(`requests/${date.getFullYear()}/${date.getMonth()}/${this.user_.uid}/`).push(game.id);
        } else {
            // Remove.
            await firebase.database().ref(`requests/${date.getFullYear()}/${date.getMonth()}/${this.user_.uid}/${key}/`).remove();
        }
    }

    /**
     * Fetches all details for a game.
     *
     * @param {BggGameData} game The game to fetch details for.
     * @return {Promise}
     */
    async fetchGameDetails(game) {
        if (game.details !== undefined) {
            return;
        }

        game.details = null;

        try {
            let games = await this.getGames();
            let response = await fetch(constants.CORS_ANYWHERE_DYNO + constants.BBG_GAME_API + game.id);
            let xml = new DOMParser().parseFromString(await response.text(), 'text/xml');

            game.details = new data.BggDetailsData(xml.children[0].children[0], games);
        } catch (e) {
            game.details = undefined; // Reset
            throw e;
        }

        return game;
    }

    /**
     * Adds a listener that gets called everytime the games change. Contains: all game requests, game requests for the current user and all games.
     *
     * @param {function(Map.<number, BggGameData>)} listener The listener.
     * @return {Promise}
     */
    async addChangeListener(listener) {
        this.changeListeners_.push(listener);

        if (this.requestedGames_ !== null) {
            listener.call(this, await this.getGames());
        }
    }

    /**
     * Removes a listener that was added using addChangeListener.
     *
     * @param {function(Map.<number, BggGameData>)} listener The previously added listener.
     */
    removeChangeListener(listener) {
        this.changeListeners_.splice(this.changeListeners_.indexOf(listener), 1);
    }

    /**
     * Get the currently logged in user.
     *
     * @return {?Object}
     */
    get user() {
        return this.user_;
    }

    /**
     * Get all users.
     *
     * @return {Map.<number, Object>}
     */
    get users() {
        return this.users_;
    }

    /***************************************************************
     * AUTHENTICATION                                              *
     ***************************************************************/

    /**
     * Logs out from firebase.
     *
     * @returns {Promise}
     */
    logout() {
        return new Promise(resolve => {
            auth.signOut().then(() => {
                this.user_ = null;
                resolve();
            });
        })
    }

    /**
     * Logs in to firebase.
     *
     * @return {Promise.<Object>}
     */
    login() {
        return new Promise((resolve, reject) => {
            auth.signInWithRedirect(provider).then(() => {
                auth.getRedirectResult().then(result => {
                    this.user_ = result.user;

                    resolve(this.user_);
                }).catch(e => {
                    console.error(e);
                    reject();
                });
            });
        });
    }

    /***************************************************************
     * PRIVATE STUFF                                               *
     ***************************************************************/

    /**
    * Wraps window.setTimeout into a promise.
    *
    * @private
    * @param {number} ms ms to sleep.
    */
    async sleep_(ms) {
        return new Promise(resolve => {
            window.setTimeout(resolve, ms);
        });
    }

    /**
     * Fetches the board game collection from boardgame geek.
     *
     * @private
     * @return {Promise.<Map.<number, BbgGameDate>>}
     */
    async fetchGames_() {
        let response;
        do {
            response = await fetch(PIETJESBAK_BBG_COLLECTION);
            if (response.status !== 200) {
                await this.sleep_(2000);
            }
        } while (response.status !== 200);

        let xml = new DOMParser().parseFromString(await response.text(), 'text/xml');

        let games = new Map();
        for (let child of xml.children[0].children) {
            let game = new data.BggGameData(child);
            games.set(game.id, game);
        }

        return games;
    }

    /**
     * Updates the user when the auth state changes.
     *
     * @private
     */
    firebaseWaitForAuthChange_() {
        auth.onAuthStateChanged(user => {
            if (user) {
                this.user_ = user;
            } else {
                this.user_ = null;
            }
        });
    }

    /**
     * Get all game requests from firebase.
     *
     * @private
     */
    async getFirebaseRequests_() {
        const date = new Date();
        firebase.database().ref(`requests/${date.getFullYear()}/${date.getMonth()}/`).on('value', async snapshot => {
            const value = snapshot.val();
            const ownUid = this.user !== null ? this.user.uid : undefined;
            this.ownRequestedGames_ = new Map();
            this.requestedGames_ = new Map();

            for (const uid in value) {
                const requests = [...new Set(Object.values(value[uid]))];
                if (uid === ownUid) {
                    const keys = Object.keys(value[uid]);
                    for (let i = 0; i < requests.length; i++) {
                        this.ownRequestedGames_.set(requests[i], keys[i]);
                    }
                }

                for (let i = 0; i < requests.length; i++) {
                    this.requestedGames_.set(requests[i], 1 + (this.requestedGames_.get(requests[i]) || 0));
                }
            }

            let games = await this.getGames();
            for (let [k, v] of games.entries()) {
                v.requestsThisMonth = this.requestedGames_.get(k) || 0;
                v.requestedByMe = this.ownRequestedGames_.has(k);
            }

            this.changeListeners_.forEach(listener => listener.call(this, games));
        });
    }

    /**
     * Get all users from firebase.
     *
     * @private
     */
    getFirebaseUsers_() {
        firebase.database().ref(`users/`).once('value', snapshot => {
            const value = snapshot.val();
            this.users_ = new Map();
            for (let uid in value) {
                this.users_.set(uid, value[uid]);
            }

            // You only have access to the users after logging in so the timing should always be right.
            if (this.user !== null && this.users_.has(this.user.uid) === false) {
                firebase.database().ref(`users/${this.user.uid}/`).set(this.user.displayName);
            }
        });
    }
}

export default new Inventory();
