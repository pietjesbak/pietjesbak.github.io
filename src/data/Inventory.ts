import { auth, database, provider } from '../Firebase';
import * as data from './BggData';
import { PIETJESBAK_BBG_COLLECTION } from './Constants';
import * as constants from './Constants';

export enum ChangeEvent {
    GAME_DATA,
    USER
}

export class Inventory {
    /**
     * A promise for all board game geek games.
     */
    private fetchGamePromise_: Promise<Map<number, data.BggGameData>> | null = null;

    /**
     * A map of games you requested and their firebase key.
     */
    private ownRequestedGames_: Map<number, string> | null = null;

    /**
     * A map of all requested games and the amount of times they were requested this month.
     */
    private requestedGames_: Map<number, number> | null = null;

    /**
     * An array of callbacks for when the game lists update.
     */
    private changeListeners_: Map<ChangeEvent, Array<(data: any) => void>> = new Map();

    /**
     * Your user object when logged in to firebase.
     */
    private user_: FirebaseUser | null = null;

    /**
     * All users of the pietjesbak. uid => name
     */
    private users_: Map<string, string> = new Map();

    /**
     * The cached firebase message.
     */
    private message_: FirebaseMessage | null = null;

    constructor() {
        this.firebaseWaitForAuthChange_();
        this.getFirebaseRequests_();
        // this.getFirebaseUsers_();
    }

    /**
     * Get the board game collection from boardgame geek.
     */
    async getGames() {
        if (this.fetchGamePromise_ === null) {
            this.fetchGamePromise_ = this.fetchGames_();
        }

        return await this.fetchGamePromise_;
    }

    /**
     * Toggles a request for a game. You need to be logged in for this.
     *
     *  @param game The game to toggle.
     */
    async toggleGame(game: data.BggGameData) {
        const date = (await this.getNextEventDate()).date;
        const key = this.ownRequestedGames_!.get(game.id);

        if (this.user_ === null) {
            throw new Error('You need to be logged in to request a game.');
        }

        if (key === undefined) {
            // Request it.
            await database.ref(`requests/${date.getFullYear()}/${date.getMonth()}/${this.user_.uid}/`).push(game.id);
        } else {
            // Remove.
            await database.ref(`requests/${date.getFullYear()}/${date.getMonth()}/${this.user_.uid}/${key}/`).remove();
        }
    }

    /**
     * Fetches all details for a game.
     *
     * @param game The game to fetch details for.
     */
    async fetchGameDetails(game: data.BggGameData) {
        if (game.details !== undefined) {
            return;
        }

        game.details = null;

        try {
            const games = await this.getGames();
            const response = await fetch(constants.CORS_ANYWHERE_DYNO + constants.BBG_GAME_API + game.id);
            const xml = new DOMParser().parseFromString(await response.text(), 'text/xml');

            game.details = new data.BggDetailsData(xml.children[0].children[0], games);
        } catch (e) {
            game.details = undefined; // Reset
            throw e;
        }

        return game;
    }

    /**
     * Adds a listener that gets called everytime the games change. Contains: all game requests, game requests for the current user and all games.
     *
     * @param event    The type of event.
     * @param listener The listener.
     */
    async addChangeListener(event: ChangeEvent, listener: (data: any) => void) {
        if (this.changeListeners_.has(event) === false) {
            this.changeListeners_.set(event, [listener]);
        } else {
            this.changeListeners_.get(event)!.push(listener);
        }

        if (event === ChangeEvent.GAME_DATA) {
            if (this.requestedGames_ !== null) {
                listener.call(this, await this.getGames());
            }
        }
    }

    /**
     * Removes a listener that was added using addChangeListener.
     *
     * @param event    The type of event.
     * @param listener The previously added listener.
     */
    removeChangeListener(event: ChangeEvent, listener: (data: any) => void) {
        if (this.changeListeners_.has(event) === true) {
            this.changeListeners_.get(event)!.splice(this.changeListeners_.get(event)!.indexOf(listener), 1);
        }
    }

    /**
     * Get the currently logged in user.
     */
    get user() {
        return this.user_;
    }

    /**
     * Get all users.
     */
    get users() {
        return this.users_;
    }

    /**
     * Get a promise containing the next event date and if this date has been confirmed.
     */
    async getNextEventDate() {
        let confirmed = true;
        let startTime;

        try {
            const event = await this.getMessage();
            startTime = new Date(event.date);
        } catch (r) {
            console.error('Failed to get next event date.');
            startTime = new Date();
        }

        const thresholdDate = new Date();
        thresholdDate.setDate(thresholdDate.getDate() + 1);

        // Adjust the start time if the event has passed.
        for (let i = 0; i < 2 && thresholdDate > startTime; i++) {
            confirmed = false;
            startTime.setMonth(thresholdDate.getMonth() + i); // Check this month and the next month for the first available event date.
            startTime.setDate(1); // Set to first day of this month.
            startTime.setDate(2 * 7 + (5 - startTime.getDay() + 7) % 7 + 1); // Figure out when the 3rd friday is from here.
        }

        return {
            confirmed,
            date: startTime
        };
    }

    /**
     * Get the message from firebase.
     */
    async getMessage(): Promise<FirebaseMessage> {
        if (this.message_ === null) {
            const message = await database.ref('message').once('value');
            const val = message.val();
            if (val === null) {
                throw new Error('Message is empty!');
            }

            this.message_ = {
                title: val.title,
                body: val.body,
                date: new Date(val.date)
            }
        }

        return this.message_;
    }

    /**
     * Update the event in firebase.
     *
     * @param title The event title.
     * @param body The event body.
     * @param date The event date.
     */
    async updateMessage(title: string, body: string, date: number) {
        await database.ref('message').set({
            title,
            body,
            date
        });

        this.message_ = null;
    }

    /***************************************************************
     * AUTHENTICATION                                              *
     ***************************************************************/

    /**
     * Logs out from firebase.
     */
    logout() {
        return new Promise(resolve => {
            auth.signOut().then(() => {
                this.user_ = null;
                resolve();
            });
        })
    }

    /**
     * Logs in to firebase.
     */
    login(): Promise<FirebaseUser> {
        return new Promise((resolve, reject) => {
            auth.signInWithRedirect(provider).then(() => {
                auth.getRedirectResult().then(result => {
                    this.user_ = result.user;

                    resolve(this.user_!);
                }).catch(e => {
                    console.error(e);
                    reject();
                });
            });
        });
    }

    /***************************************************************
     * PRIVATE STUFF                                               *
     ***************************************************************/

    /**
     * Wraps window.setTimeout into a promise.
     *
     * @param ms ms to sleep.
     */
    private async sleep_(ms: number) {
        return new Promise(resolve => {
            window.setTimeout(resolve, ms);
        });
    }

    /**
     * Fetches the board game collection from boardgame geek.
     */
    private async fetchGames_() {
        let response;
        do {
            try {
                response = await fetch(PIETJESBAK_BBG_COLLECTION);
                if (response.status !== 200) {
                    throw new Error('Collection not yet available.');
                }
            } catch (e) {
                console.log(`Trying again in 3 seconds: ${e}`);
                await this.sleep_(3000);
            }
        } while (response === undefined || response.status !== 200);

        const xml = new DOMParser().parseFromString(await response.text(), 'text/xml');

        const games = new Map();
        for (const child of xml.children[0].children as any) {
            const game = new data.BggGameData(child);
            games.set(game.id, game);
        }

        return games;
    }

    /**
     * Updates the user when the auth state changes.
     */
    private firebaseWaitForAuthChange_() {
        auth.onAuthStateChanged(async user => {
            if (user) {
                this.user_ = user as FirebaseUser;

                const admins = await database.ref('admins').once('value');
                this.user_.admin = admins.val()[user.uid] !== undefined;
            } else {
                this.user_ = null;
            }

            if (this.changeListeners_.has(ChangeEvent.USER) === true) {
                this.changeListeners_.get(ChangeEvent.USER)!.forEach(listener => listener.call(this, this.user_));
            }
        });
    }

    /**
     * Get all game requests from firebase.
     */
    private async getFirebaseRequests_() {
        const date = (await this.getNextEventDate()).date;
        database.ref(`requests/${date.getFullYear()}/${date.getMonth()}/`).on('value', async snapshot => {
            if (snapshot !== null) {
                const value = snapshot.val() as { [key: string]: number };
                const ownUid = this.user_ !== null ? this.user_.uid : undefined;
                this.ownRequestedGames_ = new Map();
                this.requestedGames_ = new Map();

                for (const uid in value) {
                    const requests = [...new Set(Object.values(value[uid]))];
                    if (uid === ownUid) {
                        const keys = Object.keys(value[uid]);
                        for (let i = 0; i < requests.length; i++) {
                            this.ownRequestedGames_.set(requests[i], keys[i]);
                        }
                    }

                    for (let i = 0; i < requests.length; i++) {
                        this.requestedGames_.set(requests[i], 1 + (this.requestedGames_.get(requests[i]) || 0));
                    }
                }

                const games = await this.getGames();
                for (const [k, v] of games.entries()) {
                    v.requestsThisMonth = this.requestedGames_.get(k) || 0;
                    v.requestedByMe = this.ownRequestedGames_.has(k);
                }

                if (this.changeListeners_.has(ChangeEvent.GAME_DATA) === true) {
                    this.changeListeners_.get(ChangeEvent.GAME_DATA)!.forEach(listener => listener.call(this, games));
                }
            }
        });
    }

    /**
     * Get all users from firebase.
     */
    // private getFirebaseUsers_() {
    //     database.ref(`users/`).once('value', snapshot => {
    //         const value = snapshot.val();
    //         this.users_ = new Map();
    //         for (const uid in value) {
    //             this.users_.set(uid, value[uid]);
    //         }

    //         // You only have access to the users after logging in so the timing should always be right.
    //         if (this.user !== null && this.users_.has(this.user.uid) === false) {
    //             database.ref(`users/${this.user.uid}/`).set(this.user.displayName);
    //         }
    //     });
    // }
}
