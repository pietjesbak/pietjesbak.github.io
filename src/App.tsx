import * as React from 'react';
import { HashRouter, Route, Switch } from 'react-router-dom';
import Tooltip from 'react-simple-tooltip';
import BggList from './BggList';
import BggRequests from './BggRequests';
import AsyncMap from './components/AsyncMap';
import inventory, { ChangeEvent } from './data/Inventory';
import EventCard from './EventCard';
import IconButton from './IconButton';
import IconLink from './IconLink';
import ScrollToTop from './ScrollToTop';

import './css/App.css';

export interface State {
    username: string;
}

class App extends React.Component<React.HTMLAttributes<App>, State> {
    constructor(props: React.HTMLAttributes<App>) {
        super(props);
        this.state = {
            username: ''
        };
    }

    componentDidMount() {
        inventory.addChangeListener(ChangeEvent.USER, user => {
            if (user) {
                this.setState({ username: (user as FirebaseUser).displayName });
            }
        });
    }

    logout = async () => {
        await inventory.logout();
        this.setState({
            username: ''
        });
    }

    login = async () => {
        const user = await inventory.login();
        this.setState({
            username: user.displayName
        });
    }

    home() {
        return (
            <div>
                <EventCard />
                <BggRequests />
            </div>
        );
    }

    games() {
        return (
            <BggList />
        );
    }

    info() {
        return (
            <div>
                <div className="card">
                    <h3>Wie zijn wij?</h3>
                    Spellenclub De pietjesbak is een spellenclub in Gent / Mariakerke. <br />
                    Elke derde vrijdag van de maand (behalve in juli en augustus) <br /><br />

                    Voor 1 € inkom is iedereen welkom om een nieuw of oud spel mee te komen spelen. <br />
                    Oud of Jong, moeilijk of makkelijk, we brengen steeds een grote variatie aan spellen mee die ter plaatse uitgelegd worden. <br />
                    We starten rond 19u30 en gaan door tot ± 23 a 24u.

                    <h3>In 't Geestje</h3>
                    <span className="location">
                        Zandloperstraat 83 <br />
                        9030 Mariakerke
                    </span>
                    <AsyncMap />
                </div>
            </div>
        );
    }

    noRoute() {
        return (
            <div className="center not-found">
                <span>404</span>
                <div style={{ fontSize: 14 + "vw", whiteSpace: "nowrap" }}>¯\_(ツ)_/¯</div>
            </div>
        )
    }

    renderLoginButton() {
        if (inventory.user !== null) {
            return <IconButton subClass="logout" icon="logout" text="Log uit" placement="left" action={this.logout} />
        } else {
            return <IconButton subClass="login" icon="login" text="Log in" placement="left" action={this.login} />
        }
    }

    render() {
        return (
            <HashRouter>
                <ScrollToTop>
                    <div className="full-height">
                        <div className="page-wrap">
                            <header>
                                <div className="wrapper">
                                    <h1>
                                        <svg id="logo" width="57" height="52">
                                            <path d="M4 25.650635094610966L16.5 4L41.5 4L54 25.650635094610966L41.5 47.30127018922193L16.5 47.30127018922193Z" strokeWidth="5" fill="#c33" stroke="#eee" />
                                        </svg>
                                        Spellenclub De Pietjesbak
                                </h1>
                                </div>
                            </header>
                            <div className="wrapper">
                                <nav className="menu">
                                    <ul>
                                        <li><IconLink icon="globe" to="/" text="Home" /></li>
                                        <li><IconLink icon="megaphone" to="/info" text="Info" /></li>
                                        <li><IconLink icon="puzzle" to="/games" text="Games" /></li>
                                        <li className="right">
                                            {this.renderLoginButton()}
                                        </li>
                                    </ul>
                                </nav>
                                {inventory.user !== null ?
                                    <div>
                                        <div className="user-profile">
                                            <Tooltip content={`Ingeloged als ${this.state.username}`} placement="left">
                                                <img alt={this.state.username} src={inventory.user.photoURL} />
                                            </Tooltip>
                                        </div>
                                    </div>
                                    :
                                    <div />
                                }
                                <Switch>
                                    <Route exact={true} path="/" component={this.home} />
                                    <Route path="/games" component={this.games} />
                                    <Route path="/info" component={this.info} />
                                    <Route render={this.noRoute} />
                                </Switch>
                            </div>
                        </div>
                        <footer>
                            <ul className="wrapper">
                                <li>
                                    <a href="https://www.facebook.com/gezelschapsspellenpietjesbak/">
                                        <i className="icon-facebook-squared" />
                                        Like
                                </a>
                                </li>
                                <li>
                                    <a href="mailto:pietdecoensel@gmail.com">
                                        <i className="icon-mail-alt" />
                                        Contact
                                </a>
                                </li>
                            </ul>
                            <span>&copy; Spellenclub De Pietjesbak &#9679; Lennert Claeys</span>
                        </footer>
                    </div>
                </ScrollToTop>
            </HashRouter>
        );
    }
}

export default App;
