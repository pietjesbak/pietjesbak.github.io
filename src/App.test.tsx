import * as enzyme from "enzyme";
import * as React from "react";

import { App } from "./App";
import { ChangeEvent, Inventory } from "./data/Inventory";

jest.mock("react-simple-tooltip", () => {
  return {
    default: "mock-tooltip",
  };
});

describe("<App />", () => {
  it("Renders the correct header", () => {
    const inventory = new Inventory() as any;

    const app = enzyme.shallow(<App inventory={inventory} />);
    expect(app.find("h1")).toMatchSnapshot();
  });

  it("Renders the correct footer", () => {
    const inventory = new Inventory() as any;

    const app = enzyme.shallow(<App inventory={inventory} />);
    expect(app.find("footer")).toMatchSnapshot();
  });

  it("Renders a logged in user", () => {
    const inventory = new Inventory() as any;
    inventory.user_ = {
      uid: "1234-5678",
      displayName: "Lennert Claeys",
      photoURL: "Photo url",
      admin: false,
    };

    const app = enzyme.shallow(<App inventory={inventory} />);

    if (inventory.changeListeners_.has(ChangeEvent.USER) === true) {
      inventory.changeListeners_
        .get(ChangeEvent.USER)!
        .forEach((listener: any) => listener.call(inventory, inventory.user_));
    }

    expect(app.find(".user-profile")).toMatchSnapshot();
    expect(app.find(".menu .right OptionIconButton button")).toMatchSnapshot();
  });

  it("Doesnt render a user", () => {
    const inventory = new Inventory() as any;
    inventory.user_ = null;

    const app = enzyme.shallow(<App inventory={inventory} />);

    if (inventory.changeListeners_.has(ChangeEvent.USER) === true) {
      inventory.changeListeners_
        .get(ChangeEvent.USER)!
        .forEach((listener: any) => listener.call(inventory, inventory.user_));
    }

    expect(app.find(".user-profile")).toHaveLength(0);
    expect(app.find(".menu .right OptionIconButton button")).toMatchSnapshot();
  });

  it("Renders a 404", () => {
    const inventory = new Inventory() as any;
    window.history.pushState({}, "404", "#/notfound/");

    const app = enzyme.mount(<App inventory={inventory} />);

    expect(app.find(".center")).toMatchSnapshot();
    app.unmount();
  });
});
