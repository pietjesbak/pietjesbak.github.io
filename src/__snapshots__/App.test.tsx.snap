// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Doesnt render a user 1`] = `
<button
  onClick={[Function]}
>
  <i
    className="icon-login"
  />
   Log in
</button>
`;

exports[`<App /> Renders a 404 1`] = `
<div
  className="center not-found"
>
  <span>
    404
  </span>
  <div
    style={
      Object {
        "fontSize": "14vw",
        "whiteSpace": "nowrap",
      }
    }
  >
    ¯\\_(ツ)_/¯
  </div>
</div>
`;

exports[`<App /> Renders a logged in user 1`] = `
<div
  className="user-profile"
>
  <mock-tooltip
    content="Ingeloged als Lennert Claeys"
    placement="left"
  >
    <img
      alt="Lennert Claeys"
      src="Photo url"
    />
  </mock-tooltip>
</div>
`;

exports[`<App /> Renders a logged in user 2`] = `
<button
  onClick={[Function]}
>
  <i
    className="icon-logout"
  />
   Log uit
</button>
`;

exports[`<App /> Renders the correct footer 1`] = `
<footer>
  <ul
    className="wrapper"
  >
    <li>
      <a
        href="https://www.facebook.com/gezelschapsspellenpietjesbak/"
      >
        <i
          className="icon-facebook-squared"
        />
        Like
      </a>
    </li>
    <li>
      <a
        href="mailto:pietdecoensel@gmail.com"
      >
        <i
          className="icon-mail-alt"
        />
        Contact
      </a>
    </li>
  </ul>
  <span>
    © Spellenclub De Pietjesbak ● Lennert Claeys
  </span>
</footer>
`;

exports[`<App /> Renders the correct header 1`] = `
<h1>
  <svg
    height="52"
    id="logo"
    width="57"
  >
    <path
      d="M4 25.650635094610966L16.5 4L41.5 4L54 25.650635094610966L41.5 47.30127018922193L16.5 47.30127018922193Z"
      fill="#c33"
      stroke="#eee"
      strokeWidth="5"
    />
  </svg>
  Spellenclub De Pietjesbak
</h1>
`;
